target/kubernetes.go

```diff
@@ -0,0 +1,173 @@
+package target
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"golang.org/x/sync/errgroup"
+
+	dto "github.com/prometheus/client_model/go"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type KubernetesEndpointFetcher struct {
+	endpointname string
+	namespace    string
+
+	port   int
+	path   string
+	scheme string
+
+	client    *http.Client
+	authToken string
+
+	kube client.Client
+
+	clock           func() time.Time
+	refreshInterval time.Duration
+	mutex           sync.Mutex
+	cache           map[string][]dto.MetricFamily
+	lastUpdated     time.Time
+}
+
+type KubernetesEndpointFetcherOpts struct {
+	Endpointname string
+	Namespace    string
+
+	Port   int
+	Path   string
+	Scheme string
+
+	AuthToken          string
+	RefreshInterval    time.Duration
+	InsecureSkipVerify bool
+}
+
+func NewKubernetesEndpointFetcher(opts KubernetesEndpointFetcherOpts) (*KubernetesEndpointFetcher, error) {
+	restConf, err := ctrl.GetConfig()
+	if err != nil {
+		return nil, err
+	}
+	kubeClient, err := client.New(restConf, client.Options{})
+	if err != nil {
+		return nil, err
+	}
+
+	return &KubernetesEndpointFetcher{
+		endpointname: opts.Endpointname,
+		namespace:    opts.Namespace,
+		port:         opts.Port,
+		path:         opts.Path,
+		scheme:       opts.Scheme,
+
+		client: &http.Client{
+			Timeout: 5 * time.Second,
+			Transport: &http.Transport{
+				TLSClientConfig: &tls.Config{
+					InsecureSkipVerify: opts.InsecureSkipVerify,
+				},
+			},
+		},
+		authToken: opts.AuthToken,
+
+		kube: kubeClient,
+
+		refreshInterval: opts.RefreshInterval,
+		mutex:           sync.Mutex{},
+		cache:           map[string][]dto.MetricFamily{},
+		lastUpdated:     time.Time{},
+	}, nil
+}
+
+func (f *KubernetesEndpointFetcher) FetchMetricsFor(ctx context.Context, endpoint string) ([]dto.MetricFamily, error) {
+	f.mutex.Lock()
```

Comment:
Maybe premature optimisation but we might want to lock the cache on a per-endpoint level.

Especially since we have very long locking locks here.

Reply:
You're right this could still be optimized a lot and more fine grained locks would probably help a lot to reduce tail latency.

However, I wanted to find a balance between low complexity and high enough performance. I would leave it as is for now and see how it performs in practice.